### ISSUE TITLE
Lost connections while scaling

### ISSUE THREAD

**Issue by caffeinism** [l7mp/stunner#151]
### Description

```mermaid
flowchart LR
  client --> web --> webrtc-server
  client <--> stunner <--> webrtc-server
  subgraph kubernetes
    web & webrtc-server & stunner
  end
```

I'm developing a media server. Users communicate by exchanging offers, ice through the webserver and relaying turns through the stunner.

I'm using this well, but I've run into a problem when scaling the pods dynamically. In a small development environment, this is fine, but when scaling up and down a large number of replicas, the stunner loses connectivity.

### Steps to Reproduce

Rolling updates to a small number of webrtc servers or scaling to a small number of webrtc servers will not cause any issues. Scaling to dozens of webrtc servers or rolling updates to a large number of webrtc servers is easily reproducible.

**Expected behavior:** Connectivity should be maintained during scale. webrtc server will gracefully terminate.

**Actual behavior:** The stunner loses the connection.

### Versions

This is reproduced in both 1.18.0 and 1.19.0.

### Info

The gateway generates logs like the one below while losing connectivity.
```
06:03:06.058863 reconcile.go:180: stunner INFO: new dataplane status: somenamespace/gateway-udp:{logLevel="all:INFO",health-check="http://:8086"}/static-auth:{realm="stunner.l7mp.io",username="<SECRET>",password="<SECRET>"}/["somenamespace/gateway-udp/turn-udp-listener":{turn://10.244.39.46:28751,public=123.123.123.123:28751,cert/key=-/-,routes=[somenamespace/somewebrtcserver]}]/["somenamespace/somewebrtcserver":{type="STATIC",protocol="UDP",endpoints=[10.244.24.52,10.244.36.88,10.244.39.59,192.168.187.184]}]/allocs:3/status=READY
06:03:06.061402 server.go:202: turn ERROR: Failed to handle datagram: unable to handle ChannelData from 10.7.72.109:1314: failed writing to socket: peer port administratively prohibited (suppressed 2 log events)
06:03:06.138467 server.go:202: turn ERROR: Failed to handle datagram: unable to handle ChannelData from 10.7.72.109:1314: failed writing to socket: peer port administratively prohibited
06:03:06.201200 server.go:202: turn ERROR: Failed to handle datagram: unable to handle ChannelData from 10.7.72.109:1314: failed writing to socket: peer port administratively prohibited
06:03:07.139439 server.go:202: turn ERROR: Failed to handle datagram: unable to handle ChannelData from 10.7.72.109:1314: failed writing to socket: peer port administratively prohibited (suppressed 12 log events)
06:03:08.138308 server.go:202: turn ERROR: Failed to handle datagram: unable to handle ChannelData from 10.7.72.109:1314: failed writing to socket: peer port administratively prohibited (suppressed 13 log events)
06:03:09.122805 server.go:202: turn ERROR: Failed to handle datagram: unable to handle ChannelData from 10.7.72.109:1314: failed writing to socket: peer port administratively prohibited (suppressed 13 log events)
06:03:10.062079 server.go:202: turn ERROR: Failed to handle datagram: unable to handle ChannelData from 10.7.72.109:1314: failed writing to socket: peer port administratively prohibited (suppressed 12 log events)
06:03:11.138560 server.go:202: turn ERROR: Failed to handle datagram: unable to handle ChannelData from 10.7.72.109:1314: failed writing to socket: peer port administratively prohibited (suppressed 14 log events)
06:03:12.123254 server.go:202: turn ERROR: Failed to handle datagram: unable to handle ChannelData from 10.7.72.109:1314: failed writing to socket: peer port administratively prohibited (suppressed 13 log events)
06:03:13.138399 server.go:202: turn ERROR: Failed to handle datagram: unable to handle ChannelData from 10.7.72.109:1314: failed writing to socket: peer port administratively prohibited (suppressed 13 log events)
06:03:14.138346 server.go:202: turn ERROR: Failed to handle datagram: unable to handle ChannelData from 10.7.72.109:1314: failed writing to socket: peer port administratively prohibited (suppressed 13 log events)
06:03:15.061727 server.go:202: turn ERROR: Failed to handle datagram: unable to handle ChannelData from 10.7.72.109:1314: failed writing to socket: peer port administratively prohibited (suppressed 12 log events)
06:03:16.138819 server.go:202: turn ERROR: Failed to handle datagram: unable to handle ChannelData from 10.7.72.109:1314: failed writing to socket: peer port administratively prohibited (suppressed 14 log events)
06:03:17.062617 server.go:202: turn ERROR: Failed to handle datagram: unable to handle ChannelData from 10.7.72.109:1314: failed writing to socket: peer port administratively prohibited (suppressed 12 log events)
```

#### Operator logs 

Basically, I followed [this tutoral](https://github.com/l7mp/stunner/blob/main/docs/examples/kurento-one2one-call/kurento-one2one-call-stunner.yaml) and there is nothing unusual about the operator and api status.

**Comment by rg0now** [comment]
This is not a STUNner issue per se. We've seen this problem several times and the reason is the way your load-balancer ingests UDP traffic into the cluster. Unfortunately some K8s load-balancers do not create conntrack state for UDP connections (which would make sure connections are pinned to a specific pod) but rather use a dumb stateless hash function to assign a specific STUNner pod for each UDP packet separatly. When you add/remove pods, hash-ranges change and connections get rerouted. Some LBs use [consistent hashing](https://medium.com/swlh/load-balancing-and-consistent-hashing-5fe0156035e1), but that is still amenable to rare connections losses during scaling.

I'm afraid the only solution is to use TCP instead of UDP (this is suboptimal but TCP connections are always lod-balanced using per-connection conntrack state) or migrate to another k8s provider and/or LB that implements a smarter UDP ingress. May I ask how you're hosting your testbed?

**Comment by caffeinism** [comment]
I'm using AKS. Is there any documentation that summarizes the load balancer compatibility of hosted Kubernetes services?

**Comment by rg0now** [comment]
Thanks. We're currently compiling a compatibility matrix, hope to release it soon. The last time we checked the GKE UDP LB was working fine.

Closing this for now, feel free to reopen if anything comes up.

**Comment by caffeinism** [comment]
@rg0now But I'm a little curious, I didn't scale the gateway (1 replicas), just the server that webrtc communicates with. Is this also affected by the loadbalancer, even though there is only 1 pod that the loadbalancer is connected to?

**Comment by rg0now** [comment]
Oh, I"m terribly sorry I completely misunderstood, thanks for clarifying. If it is scale-down that breaks then maybe you're hit by this [issue](https://github.com/l7mp/stunner/issues/138). This problem has been fixed [here](https://github.com/l7mp/stunner-gateway-operator/commit/d7c0a10118617049caa3732bc8fb8ca50e376c1f) but you must run a fairly recent dev release to have that fix. If this affects scale-up, can you specify the steps to reproduce this?

**Comment by caffeinism** [comment]
I changed the image tags for stunnerd and stunner operator to dev, but the error seems to happen the same. 

I don't see anything particularly strange in the gateway logs when scaling up. (This may not be an issue with the stunner).

**Comment by rg0now** [comment]
I tried my best to reproduce this, to no avail. 

The offending error log indicates that your client is trying to send to a backend IP that is no longer part of the backend Service (the one in the UDPRoute backendRefs). This may happen if you scale the corresponding backend pod down (don't do that) but forget to tell the client to do an ICE restart (try to avoid this). Scaling webrtc servers is difficult, see some background [here](https://docs.livekit.io/home/self-hosting/kubernetes/#Graceful-restarts).

I am going to close this in preparation for the v1 release. Feel free to reopen if you find a simple way to reproduce this. We have had some issues with STUNner graceful shutdown and it would be critical we get this right...
